apps/bot_manager/lib/bot_state_machine.ex

```diff
@@ -3,63 +3,233 @@ defmodule BotManager.BotStateMachine do
   This module will take care of deciding what the bot will do on each deciding step
   """
 
+  alias BotManager.BotStateMachineChecker
   alias BotManager.Utils
   alias BotManager.Math.Vector
+  require Logger
 
-  def decide_action(%{bots_enabled?: false}) do
-    {:move, %{x: 0, y: 0}}
+  @skill_1_key "1"
+  @skill_2_key "2"
+  @dash_skill_key "3"
+  @vision_range 1200
+
+  def decide_action(%{bots_enabled?: false, bot_state_machine: bot_state_machine}) do
+    %{action: {:move, %{x: 0, y: 0}}, bot_state_machine: bot_state_machine}
   end
 
-  def decide_action(%{game_state: game_state, bot_player: bot_player, attack_blocked: attack_blocked}) do
-    closest_player =
-      map_directions_to_players(game_state, bot_player)
-      |> Enum.min_by(fn player_info -> player_info.distance end)
+  def decide_action(%{
+        game_state: game_state,
+        bot_player: bot_player,
+        bot_state_machine: bot_state_machine,
+        attack_blocked: attack_blocked
+      }) do
+    bot_state_machine =
+      if is_nil(bot_state_machine.previous_position) do
+        bot_state_machine
+        |> Map.put(:previous_position, bot_player.position)
+        |> Map.put(:current_position, bot_player.position)
+      else
+        bot_state_machine
+        |> Map.put(:previous_position, bot_state_machine.current_position)
+        |> Map.put(:current_position, bot_player.position)
+      end
 
-    random_distance = 1000
+    %{distance: distance} =
+      get_distance_and_direction_to_positions(bot_state_machine.previous_position, bot_state_machine.current_position)
 
-    cond do
-      attack_blocked ->
-        :stand
+    bot_state_machine =
+      Map.put(bot_state_machine, :progress_for_basic_skill, bot_state_machine.progress_for_basic_skill + distance)
+
+    next_state = BotStateMachineChecker.move_to_next_state(bot_player, bot_state_machine)
 
-      closest_player.distance > random_distance ->
-        {:move, closest_player.direction}
+    case next_state do
+      :moving ->
+        direction = maybe_switch_direction(bot_player, bot_state_machine)
 
-      closest_player.distance < 50 ->
-        {:move, create_random_direction()}
+        %{
+          action: determine_player_move_action(bot_player, bot_state_machine, direction),
+          bot_state_machine: bot_state_machine
+        }
 
-      closest_player.distance <= random_distance ->
-        {:attack, closest_player.direction}
+      :aggresive ->
+        use_skill(%{
+          bot_player: bot_player,
+          bot_state_machine: bot_state_machine,
+          game_state: game_state,
+          attack_blocked: attack_blocked
+        })
 
-      true ->
-        {:move, create_random_direction()}
+      :running_away ->
+        ran_away(bot_player, game_state, bot_state_machine)
```

Comment:
isn't this run_away? this function makes the bot run in the opposite direction, right?

Reply:
yes! I'll rename it